File: backend/api/v1/serializers/settlements.py
from __future__ import annotations

from rest_framework import serializers
from settlements.models import (
    BankStatementImport,
    BankStatementLine,
    CashierDayClose,
    SettlementAllocation,
    SettlementAllocationSourceType,
    UpiSettlementImport,
    UpiSettlementLine,
)

class BankStatementLineSerializer(serializers.ModelSerializer):
    class Meta:
        model = BankStatementLine
        fields = [
            "id",
            "statement_import",
            "transaction_date",
            "value_date",
            "description",
            "reference_no",
            "debit",
            "credit",
            "balance",
            "raw_payload",
            "normalized_reference",
            "matched_status",
            "metadata",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields


class BankStatementImportSerializer(serializers.ModelSerializer):
    uploaded_by_username = serializers.CharField(source="uploaded_by.username", read_only=True)
    bank_finance_account_name = serializers.CharField(source="bank_finance_account.name", read_only=True)

    class Meta:
        model = BankStatementImport
        fields = [
            "id",
            "import_no",
            "bank_finance_account",
            "bank_finance_account_name",
            "statement_period_from",
            "statement_period_to",
            "uploaded_file",
            "uploaded_by",
            "uploaded_by_username",
            "uploaded_at",
            "status",
            "checksum",
            "metadata",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "import_no",
            "uploaded_by",
            "uploaded_at",
            "status",
            "checksum",
            "metadata",
            "created_at",
            "updated_at",
        ]


class BankStatementImportCreateSerializer(serializers.ModelSerializer):
    class Meta:
        model = BankStatementImport
        fields = [
            "id",
            "bank_finance_account",
            "statement_period_from",
            "statement_period_to",
            "uploaded_file",
        ]

    def validate(self, attrs):
        attrs = super().validate(attrs)
        p_from = attrs.get("statement_period_from")
        p_to = attrs.get("statement_period_to")
        if p_from and p_to and p_to < p_from:
            raise serializers.ValidationError(
                {"statement_period_to": "Statement period end cannot be earlier than start."}
            )
        file = attrs.get("uploaded_file")
        if not file:
            raise serializers.ValidationError({"uploaded_file": "File is required."})
        return attrs


class UpiSettlementLineSerializer(serializers.ModelSerializer):
    class Meta:
        model = UpiSettlementLine
        fields = [
            "id",
            "settlement_import",
            "transaction_ref",
            "payment_ref",
            "gross_amount",
            "fee_amount",
            "net_amount",
            "settlement_date",
            "raw_payload",
            "matched_status",
            "metadata",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields


class UpiSettlementImportSerializer(serializers.ModelSerializer):
    uploaded_by_username = serializers.CharField(source="uploaded_by.username", read_only=True)
    upi_finance_account_name = serializers.CharField(source="upi_finance_account.name", read_only=True)

    class Meta:
        model = UpiSettlementImport
        fields = [
            "id",
            "import_no",
            "upi_finance_account",
            "upi_finance_account_name",
            "settlement_date",
            "uploaded_file",
            "uploaded_by",
            "uploaded_by_username",
            "uploaded_at",
            "status",
            "checksum",
            "metadata",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "import_no",
            "uploaded_by",
            "uploaded_at",
            "status",
            "checksum",
            "metadata",
            "created_at",
            "updated_at",
        ]


class UpiSettlementImportCreateSerializer(serializers.ModelSerializer):
    class Meta:
        model = UpiSettlementImport
        fields = [
            "id",
            "upi_finance_account",
            "settlement_date",
            "uploaded_file",
        ]

    def validate(self, attrs):
        attrs = super().validate(attrs)
        file = attrs.get("uploaded_file")
        if not file:
            raise serializers.ValidationError({"uploaded_file": "File is required."})
        return attrs


class SettlementAllocationSerializer(serializers.ModelSerializer):
    matched_by_username = serializers.CharField(source="matched_by.username", read_only=True)
    finance_account_name = serializers.CharField(source="finance_account.name", read_only=True)

    class Meta:
        model = SettlementAllocation
        fields = [
            "id",
            "source_type",
            "source_id",
            "finance_account",
            "finance_account_name",
            "matched_amount",
            "status",
            "payment",
            "receipt",
            "money_movement",
            "matched_by",
            "matched_by_username",
            "matched_at",
            "confidence",
            "metadata",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields


class SettlementAllocationCreateSerializer(serializers.Serializer):
    source_type = serializers.ChoiceField(choices=SettlementAllocationSourceType.choices)
    source_id = serializers.CharField()
    finance_account = serializers.IntegerField()
    matched_amount = serializers.DecimalField(max_digits=12, decimal_places=2)
    payment = serializers.IntegerField(required=False, allow_null=True)
    receipt = serializers.IntegerField(required=False, allow_null=True)
    money_movement = serializers.IntegerField(required=False, allow_null=True)
    note = serializers.CharField(required=False, allow_blank=True, default="")

    def validate(self, attrs):
        attrs = super().validate(attrs)
        if not (attrs.get("payment") or attrs.get("receipt") or attrs.get("money_movement")):
            raise serializers.ValidationError({"payment": "At least one target (payment, receipt, money_movement) is required."})
        if attrs.get("matched_amount") is not None and attrs["matched_amount"] <= 0:
            raise serializers.ValidationError({"matched_amount": "matched_amount must be greater than zero."})
        return attrs


class SettlementAllocationVoidSerializer(serializers.Serializer):
    reason = serializers.CharField(required=False, allow_blank=True, default="")


# === Cashier Day Close Serializers ===


class CashierDayCloseSerializer(serializers.ModelSerializer):
    """Read-only detail serializer for cashier day-close records."""
    cashier_username = serializers.CharField(source="cashier.username", read_only=True)
    branch_code = serializers.CharField(source="branch.code", read_only=True, allow_null=True)
    branch_name = serializers.CharField(source="branch.name", read_only=True, allow_null=True)
    cash_counter_name = serializers.CharField(source="cash_counter.name", read_only=True, allow_null=True)
    finance_account_name = serializers.CharField(source="finance_account.name", read_only=True, allow_null=True)
    closed_by_username = serializers.CharField(source="closed_by.username", read_only=True, allow_null=True)
    approved_by_username = serializers.CharField(source="approved_by.username", read_only=True, allow_null=True)

    class Meta:
        model = CashierDayClose
        fields = [
            "id",
            "close_no",
            "cashier",
            "cashier_username",
            "branch",
            "branch_code",
            "branch_name",
            "cash_counter",
            "cash_counter_name",
            "finance_account",
            "finance_account_name",
            "business_date",
            "opening_cash",
            "system_cash_total",
            "counted_cash",
            "variance",
            "status",
            "closed_by",
            "closed_by_username",
            "closed_at",
            "approved_by",
            "approved_by_username",
            "approved_at",
            "notes",
            "metadata",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields


class CashierDayCloseCreateSerializer(serializers.Serializer):
    """Serializer for creating a new cashier day-close draft."""
    business_date = serializers.DateField()
    counted_cash = serializers.DecimalField(max_digits=12, decimal_places=2)
    branch = serializers.IntegerField(required=False, allow_null=True)
    cash_counter = serializers.IntegerField(required=False, allow_null=True)
    finance_account = serializers.IntegerField(required=False, allow_null=True)
    opening_cash = serializers.DecimalField(max_digits=12, decimal_places=2, required=False, default="0.00")
    notes = serializers.CharField(required=False, allow_blank=True, default="")

    def validate(self, attrs):
        attrs = super().validate(attrs)
        if attrs.get("counted_cash") is not None and attrs["counted_cash"] < 0:
            raise serializers.ValidationError({"counted_cash": "Counted cash cannot be negative."})
        return attrs


class CashierDayCloseSubmitSerializer(serializers.Serializer):
    """Serializer for submitting a cashier day-close (DRAFT → SUBMITTED)."""
    # No input fields; just accept the submission request


class CashierDayCloseApprovalSerializer(serializers.Serializer):
    """Serializer for admin approval of a day-close (SUBMITTED → APPROVED)."""
    notes = serializers.CharField(required=False, allow_blank=True, default="")


class CashierDayCloseRejectSerializer(serializers.Serializer):
    """Serializer for admin rejection of a day-close (SUBMITTED → REJECTED)."""
    notes = serializers.CharField(required=True, allow_blank=False)

    def validate_notes(self, value):
        if not value or not value.strip():
            raise serializers.ValidationError("Rejection notes are required and cannot be empty.")
        return value.strip()

File: frontend/src/types/settlements.ts
export type PaginatedResponse<T> = {
  count: number;
  next: string | null;
  previous: string | null;
  results: T[];
};

export type SettlementSourceType =
  | "BANK_STATEMENT_LINE"
  | "UPI_SETTLEMENT_LINE"
  | "CASHIER_DAY_CLOSE";

export type BankStatementImport = {
  id: number;
  import_no: string;
  bank_finance_account: number;
  bank_finance_account_name?: string;
  statement_period_from: string;
  statement_period_to: string;
  uploaded_file?: string;
  uploaded_by: number;
  uploaded_by_username?: string;
  uploaded_at: string;
  status: string;
  checksum: string;
  metadata?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type BankStatementLine = {
  id: number;
  statement_import: number;
  transaction_date: string;
  value_date?: string | null;
  description: string;
  reference_no?: string | null;
  debit: string;
  credit: string;
  balance?: string | null;
  raw_payload?: Record<string, unknown> | null;
  normalized_reference?: string | null;
  matched_status: string;
  metadata?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type UpiSettlementImport = {
  id: number;
  import_no: string;
  upi_finance_account: number;
  upi_finance_account_name?: string;
  settlement_date: string;
  uploaded_file?: string;
  uploaded_by: number;
  uploaded_by_username?: string;
  uploaded_at: string;
  status: string;
  checksum: string;
  metadata?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type UpiSettlementLine = {
  id: number;
  settlement_import: number;
  transaction_ref: string;
  payment_ref?: string | null;
  gross_amount: string;
  fee_amount: string;
  net_amount: string;
  settlement_date: string;
  raw_payload?: Record<string, unknown> | null;
  matched_status: string;
  metadata?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type SettlementAllocation = {
  id: number;
  source_type: SettlementSourceType;
  source_id: string;
  finance_account: number;
  finance_account_name?: string;
  matched_amount: string;
  status: string;
  payment?: number | null;
  receipt?: number | null;
  money_movement?: number | null;
  matched_by?: number | null;
  matched_by_username?: string;
  matched_at?: string | null;
  confidence?: string | null;
  metadata?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type BankImportCreatePayload = {
  bank_finance_account: number;
  statement_period_from: string;
  statement_period_to: string;
  uploaded_file: File;
};

export type UpiImportCreatePayload = {
  upi_finance_account: number;
  settlement_date: string;
  uploaded_file: File;
};

export type SettlementAllocationCreatePayload = {
  source_type: SettlementSourceType;
  source_id: string;
  finance_account: number;
  matched_amount: string;
  payment?: number | null;
  receipt?: number | null;
  money_movement?: number | null;
  note?: string;
};

export type SettlementAllocationVoidPayload = {
  reason?: string;
};

// === Cashier Day Close Types ===

export type CashierDayCloseStatus = 
  | "DRAFT"
  | "SUBMITTED"
  | "APPROVED"
  | "REJECTED"
  | "VOIDED";

export type CashierDayClose = {
  id: number;
  close_no: string;
  cashier: number;
  cashier_username?: string;
  branch?: number | null;
  branch_code?: string | null;
  branch_name?: string | null;
  cash_counter?: number | null;
  cash_counter_name?: string | null;
  finance_account?: number | null;
  finance_account_name?: string | null;
  business_date: string;
  opening_cash: string;
  system_cash_total: string;
  counted_cash: string;
  variance: string;
  status: CashierDayCloseStatus;
  closed_by?: number | null;
  closed_by_username?: string | null;
  closed_at?: string | null;
  approved_by?: number | null;
  approved_by_username?: string | null;
  approved_at?: string | null;
  notes: string;
  metadata?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type CashierDayCloseCreatePayload = {
  business_date: string;
  counted_cash: string;
  branch?: number | null;
  cash_counter?: number | null;
  finance_account?: number | null;
  opening_cash?: string;
  notes?: string;
};

export type CashierDayClosePreviewResponse = {
  business_date: string;
  system_cash_total: string;
};

export type CashierDayCloseApprovalPayload = {
  notes?: string;
};

export type CashierDayCloseRejectPayload = {
  notes: string;
};
